cmd/wallet-js-sdk/src/oidc/issuance/openid4ci.js

```diff
@@ -1,212 +1,287 @@
- * @param {Object} options
+ * @param {string} callback_uri
+ * @param {string} client_state
+ * @param {Object} client_config
+ * @param {string} client_config.user_did - the DID of the wallet user to whom the credential is being issued.
+ * @param {string} client_config.wallet_callback_uri - wallet-hosted URI for callback redirect from issuer, after user authorizes issuer.
+ * @param {string} client_config.client_id - the wallet app instance's OIDC client ID. * @param {Object} client_state
  * @param {function} jwt_signer - a handler for signing a JWT for proving possession of user_did.
  *    When this creates a JWT, it should add kid, jwk, or x5c field corresponding to a key bound to user_did.
- * @param {function} load_transaction - a handler for loading an OIDC issuance transaction under the given transaction ID.
- *    Used for loading transaction state saved by requestIssuance.
- * @param {function} delete_transaction - a handler for deleting an OIDC issuance transaction under the given transaction ID.
  */
-export async function CompleteIssuance(
-  oidc_state = '',
-  user_did = '',
-  {
-    authorization_code = '',
-    pre_authorized_code = '',
-    user_pin = '',
-  },
-  jwt_signer = async function(iss, aud, iat, c_nonce){return {}},
-  // TODO: should these handlers be parameterless instead, as caller-provided closures on the txn ID?
-  //  then the caller doesn't need to trust that this API will only touch the relevant transaction.
-  load_transaction = async function(txn_id= ''){return {}},
-  delete_transaction = async function(txn_id = ''){},
+export async function callback(
+  callback_uri,
+  client_state,
+  client_config,
+  jwt_signer = async function(iss, aud, iat, c_nonce){return ''},
 ) {
+  const parsedURI = new URL(callback_uri);
 
-  const transaction_data = await load_transaction(oidc_state);
+  const authorization_code = parsedURI.searchParams.get('code');
+  const oauth_state = parsedURI.searchParams.get('state');
 
-  // stub
-  const exchangeAuthCode = async function(auth_code, token_endpoint) {
-    // data returned by issuer server
-    return {
-      access_token: '',
-      token_type: '',
-      c_nonce: '',
-    };
-  };
+  if (authorization_code === '') {
+    throw new Error('Callback URI is missing authorization `code` as a query parameter.');
+  }
 
-  // stub
-  const exchangePreAuthCode = async function(pre_auth_code, user_pin, token_endpoint) {
-    // data returned by issuer server
-    return {
-      access_token: '',
-      token_type: '',
-      c_nonce: '',
-      authorization_pending: true,
-      interval: 1,
-    };
-  };
+  if (oauth_state === '') {
+    throw new Error('Issuer did not include state parameter in callback URI when returning user to wallet.')
+  }
+
+  if (client_state === '') {
+    throw new Error('Cannot complete authorization flow, `callback` must be provided with client_state string returned by ' +
+      'previous `authorize` call.')
+  }
```

Comment:
Since you don't seem to need the url itself, why not use the `URLSearchParams` instead?

Another thing is that it returns `null` if the parameter does not exist, so you'd need to check for that instead:

Reply:
done https://github.com/trustbloc/agent-sdk/pull/452/files#diff-b6fc870019a039dfefb1922ebe309e51614712b835bd26d6b2a6d79c09fb8aa1R130